### PR TITLE
[MSGINA] Convert shutdown and logoff boxes back to modal dialog boxes

### PR DESCRIPTION
**Pull request by arnavbhatt288**
## Purpose

Convert back to modal dialog boxes to fix bugs. Taken from [this](https://github.com/reactos/reactos/pull/3999) closed pull request.

JIRA issue: [CORE-17749](https://jira.reactos.org/browse/CORE-17749)

## Proposed changes

- Convert all modeless dialog boxes to dialog boxes.
- Use WM_TIMER to detect `SHIFT` key presses instead.
- Remove some useless functions.

### CODE REVIEW

File: dll/win32/msgina/shutdown.c [hunk 1]
```diff
@@ -848,140 +872,141 @@ ShutdownOnInit(
...
+                                pContext->pgContext->hDllInstance);
+
+        /* Gather old button func */
+        pContext->OldButtonProc = (WNDPROC)GetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_HIBERNATE), GWLP_WNDPROC);
+
+        /* Make buttons to remember pContext and subclass the buttons */
+        for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
+        {
+            SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_HIBERNATE + i), GWLP_USERDATA, (LONG_PTR)pContext);
+            SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_HIBERNATE + i), GWLP_WNDPROC, (LONG_PTR)OwnerDrawButtonSubclass);
```
Comment (learn-more) [inline dll/win32/msgina/shutdown.c:927]: Why not use `SetWindowSubclass` instead?

File: dll/win32/msgina/shutdown.c [hunk 2]
```diff
@@ -848,140 +872,141 @@ ShutdownOnInit(
...
+        if (pContext->ShutdownOptions & WLX_SHUTDOWN_STATE_REBOOT)
+        {
+            LoadStringW(pgContext->hDllInstance, IDS_SHUTDOWN_RESTART, szBuffer, _countof(szBuffer));
+            idx = SendMessageW(hwndList, CB_ADDSTRING, 0, (LPARAM)szBuffer);
+            if (idx != CB_ERR)
+                SendMessageW(hwndList, CB_SETITEMDATA, idx, WLX_SAS_ACTION_SHUTDOWN_REBOOT);
+        }
 
-    // if (pContext->ShutdownOptions & 0x20) {}
+        // if (pContext->ShutdownOptions & 0x08) {}
```
Comment (learn-more) [inline dll/win32/msgina/shutdown.c:972]: What purpose do these comments serve?

File: dll/win32/msgina/shutdown.c [hunk 3]
```diff
@@ -678,31 +699,34 @@ ReplaceRequiredButton(
...
+        SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
+    }
+    InvalidateRect(hDlg, NULL, FALSE);
+}
+
+VOID OnTimer(
+    HWND hDlg,
+    PSHUTDOWN_DLG_CONTEXT pContext)
+{
+    BOOL ReplaceButton = !!(GetKeyState(VK_SHIFT) & 0x8000);
```
Comment (learn-more) [inline dll/win32/msgina/shutdown.c:719]: Why are you using a timer for this?

**Review by learn-more — CHANGES_REQUESTED** [review]

File: dll/win32/msgina/shutdown.c (hunk 2 above)
Comment (arnavbhatt288) [reply dll/win32/msgina/shutdown.c:972]: These comments were existed before my changes, hence I thought to keep it as it is.

File: dll/win32/msgina/shutdown.c (hunk 1 above)
Comment (arnavbhatt288) [reply dll/win32/msgina/shutdown.c:927]: MSGINA shouldn't depend on components like Comctl32 to work, which `SetWindowSubclass` depends upon.

File: dll/win32/msgina/shutdown.c [hunk 4]
```diff
@@ -848,140 +872,141 @@ ShutdownOnInit(
...
+                                hDlg, IDS_SHUTDOWN_SHUTDOWN,
+                            pContext->pgContext->hDllInstance);
+        CreateToolTipForButtons(IDC_BUTTON_REBOOT,
+                                IDS_SHUTDOWN_RESTART_DESC,
+                                hDlg, IDS_SHUTDOWN_RESTART,
+                                pContext->pgContext->hDllInstance);
+        CreateToolTipForButtons(IDC_BUTTON_SLEEP,
+                                IDS_SHUTDOWN_SLEEP_DESC,
+                                hDlg, IDS_SHUTDOWN_SLEEP,
+                                pContext->pgContext->hDllInstance);
```
[the remark this answers is not in the document]
Comment (arnavbhatt288) [reply dll/win32/msgina/shutdown.c:918]: I already have a `for loop` which does the similar tasks like subclassing those buttons or making it to remember a structure. I am not sure why I haven't done for `CreateToolTipForButtons` as well from the get go.

File: dll/win32/msgina/shutdown.c [hunk 5]
```diff
@@ -662,7 +684,7 @@ ReplaceRequiredButton(
     /* Get the position of the destination button */
     GetWindowRect(hwndDest, &rect);
 
-    /* Get the corrected translated coordinates which is relative to the client window */
+    /* Get the corrected translated coordinates which is relative to the client window */
```
Comment (binarymaster) [inline dll/win32/msgina/shutdown.c:687]:
```suggestion
    /* Get the corrected translated coordinates which is relative to the client window */
```

File: dll/win32/msgina/shutdown.c [hunk 6]
```diff
@@ -678,31 +700,34 @@ ReplaceRequiredButton(
...
-        }
-        else
-        {
-            LoadStringW(hInstance, IDS_SHUTDOWN_SLEEP, szBuffer, _countof(szBuffer));
-            SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
-        }
+        LoadStringW(pContext->pgContext->hDllInstance, IDS_SHUTDOWN_SLEEP, szBuffer, _countof(szBuffer));
+        SetDlgItemTextW(hDlg, IDC_SLEEP_STATIC, szBuffer);
+    }
+    InvalidateRect(hDlg, NULL, FALSE);
```
Comment (binarymaster) [inline dll/win32/msgina/shutdown.c:713]:
```suggestion

    InvalidateRect(hDlg, NULL, FALSE);
```

File: dll/win32/msgina/shutdown.c [hunk 7]
```diff
@@ -637,11 +640,30 @@ CreateToolTipForButtons(
...
 VOID
-ReplaceRequiredButton(
+EndFriendlyDialog(
+    HWND hDlg,
+    PSHUTDOWN_DLG_CONTEXT pContext)
+{
+    if (pContext->bTimer)
+    {
+        KillTimer(hDlg, pContext->iTimer);
+    }
```
Comment (binarymaster) [inline dll/win32/msgina/shutdown.c:650]:
```suggestion
    }

```

File: dll/win32/msgina/shutdown.c [hunk 8]
```diff
@@ -848,140 +873,122 @@ ShutdownOnInit(
...
+
+        /* Gather image strip */
+        pContext->hImageStrip = LoadBitmapW(pgContext->hDllInstance, MAKEINTRESOURCEW(IDB_IMAGE_STRIP));
+        
+        /* Set the boolean flags to false */
+        pContext->bIsSleepButtonReplaced = FALSE;
+        pContext->bTimer = FALSE;
+
+       EnableWindow(GetDlgItem(hDlg, IDC_BUTTON_HIBERNATE), FALSE);
+       EnableWindow(GetDlgItem(hDlg, IDC_BUTTON_SLEEP), IsPwrSuspendAllowed());
```
Comment (binarymaster) [inline dll/win32/msgina/shutdown.c:896]:
```suggestion
        EnableWindow(GetDlgItem(hDlg, IDC_BUTTON_HIBERNATE), FALSE);
        EnableWindow(GetDlgItem(hDlg, IDC_BUTTON_SLEEP), IsPwrSuspendAllowed());
```

File: dll/win32/shell32/dialogs/dialogs.cpp [hunk 9]
```diff
@@ -1389,18 +1418,15 @@ static VOID FancyLogoffOnInit(HWND hwnd, PLOGOFF_DLG_CONTEXT pContext)
...
     pContext->OldButtonProc = (WNDPROC)GetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON), GWLP_WNDPROC);
 
-    /* Make buttons to remember pContext and subclass the buttons as well as set bIsButtonHot boolean flags to false */
+    /* Set bIsButtonHot to false, create tooltips for each buttons and subclass the buttons */
     for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
     {
         pContext->bIsButtonHot[i] = FALSE;
-        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_USERDATA, (LONG_PTR)pContext);
         SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_WNDPROC, (LONG_PTR)OwnerDrawButtonSubclass);
+        CreateToolTipForButtons(IDC_LOG_OFF_BUTTON + i, IDS_LOG_OFF_DESC + i, hwnd, IDS_LOG_OFF_TITLE + i);
```
Comment (binarymaster) [inline dll/win32/shell32/dialogs/dialogs.cpp:1429]:
```suggestion
        CreateToolTipForButtons(IDC_LOG_OFF_BUTTON + i,
                                IDS_LOG_OFF_DESC + i,
                                hwnd,
                                IDS_LOG_OFF_TITLE + i);
```

File: dll/win32/shell32/dialogs/dialogs.cpp [hunk 10]
```diff
@@ -1539,36 +1551,7 @@ EXTERN_C int WINAPI LogoffWindowsDialog(HWND hWndOwner)
...
-                        GetDlgItemTextW(hWndChild, IDC_LOG_OFF_BUTTON + i, szBuffer, _countof(szBuffer));
-                        SetDlgItemTextW(hWndChild, IDC_LOG_OFF_STATIC + i, szBuffer);
-                    }
-                    bIsAltKeyPressed = TRUE;
-                }
-            }
-            break;
-        }
-    }
+    DialogBoxParamW(shell32_hInstance, MAKEINTRESOURCEW(LogoffDialogID), parent, LogOffDialogProc, (LPARAM)&Context);
```
Comment (binarymaster) [inline dll/win32/shell32/dialogs/dialogs.cpp:1554]:
```suggestion
    DialogBoxParamW(shell32_hInstance,
                    MAKEINTRESOURCEW(LogoffDialogID),
                    parent,
                    LogOffDialogProc,
                    (LPARAM)&Context);
```

File: dll/win32/msgina/shutdown.c [hunk 11]
```diff
@@ -637,11 +640,30 @@ CreateToolTipForButtons(
...
+    }   
+    DeleteObject(pContext->hBitmap);
+    DeleteObject(pContext->hBrush);
+    DeleteObject(pContext->hImageStrip);
+    DeleteObject(pContext->hfFont);
+
+    /* Remove the subclass from the buttons */
+    for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
+    {
+        SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_SHUTDOWN + i), GWLP_WNDPROC, (LONG_PTR)pContext->OldButtonProc);
```
Comment (binarymaster) [inline dll/win32/msgina/shutdown.c:659]:
```suggestion
        SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_SHUTDOWN + i),
                          GWLP_WNDPROC,
                          (LONG_PTR)pContext->OldButtonProc);
```

File: dll/win32/msgina/shutdown.c [hunk 12]
```diff
@@ -848,140 +873,122 @@ ShutdownOnInit(
...
+       EnableWindow(GetDlgItem(hDlg, IDC_BUTTON_SLEEP), IsPwrSuspendAllowed());
+
+        /* Gather old button func */
+        pContext->OldButtonProc = (WNDPROC)GetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_HIBERNATE), GWLP_WNDPROC);
+
+        /* Set bIsButtonHot to false, create tooltips for each buttons, make buttons to remember pContext and subclass the buttons */
+        for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
+        {
+            pContext->bIsButtonHot[i] = FALSE;
+            SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_SHUTDOWN + i), GWLP_WNDPROC, (LONG_PTR)OwnerDrawButtonSubclass);
```
Comment (binarymaster) [inline dll/win32/msgina/shutdown.c:905]:
```suggestion
            SetWindowLongPtrW(GetDlgItem(hDlg, IDC_BUTTON_SHUTDOWN + i),
                              GWLP_WNDPROC,
                              (LONG_PTR)OwnerDrawButtonSubclass);
```

File: dll/win32/shell32/dialogs/dialogs.cpp [hunk 13]
```diff
@@ -1300,6 +1316,19 @@ VOID CreateToolTipForButtons(int controlID, int detailID, HWND hDlg, int titleID
...
+VOID EndFriendlyDialog(HWND hwnd, PLOGOFF_DLG_CONTEXT pContext)
+{
+    DeleteObject(pContext->hBrush);
+    DeleteObject(pContext->hImageStrip);
+    DeleteObject(pContext->hfFont);
+
+    /* Remove the subclass from the buttons */
+    for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
+    {
+        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_WNDPROC, (LONG_PTR)pContext->OldButtonProc);
```
Comment (binarymaster) [inline dll/win32/shell32/dialogs/dialogs.cpp:1328]:
```suggestion
        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i),
                          GWLP_WNDPROC,
                          (LONG_PTR)pContext->OldButtonProc);
```

File: dll/win32/shell32/dialogs/dialogs.cpp [hunk 14]
```diff
@@ -1389,18 +1418,15 @@ static VOID FancyLogoffOnInit(HWND hwnd, PLOGOFF_DLG_CONTEXT pContext)
...
     /* Gather old button func */
     pContext->OldButtonProc = (WNDPROC)GetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON), GWLP_WNDPROC);
 
-    /* Make buttons to remember pContext and subclass the buttons as well as set bIsButtonHot boolean flags to false */
+    /* Set bIsButtonHot to false, create tooltips for each buttons and subclass the buttons */
     for (int i = 0; i < NUMBER_OF_BUTTONS; i++)
     {
         pContext->bIsButtonHot[i] = FALSE;
-        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_USERDATA, (LONG_PTR)pContext);
         SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i), GWLP_WNDPROC, (LONG_PTR)OwnerDrawButtonSubclass);
```
Comment (binarymaster) [inline dll/win32/shell32/dialogs/dialogs.cpp:1428]:
```suggestion
        SetWindowLongPtrW(GetDlgItem(hwnd, IDC_LOG_OFF_BUTTON + i),
                          GWLP_WNDPROC,
                          (LONG_PTR)OwnerDrawButtonSubclass);
```

**Review by binarymaster — COMMENTED** [review]
Coding style stuff (will apply it myself)

**Review by binarymaster — APPROVED** [review]
This PR fixes the bug report for me, I'm going to merge it in a few days if noone propose a better fix. Two notes though:
- I have reduced the diff in this PR for the sake of simplicity
- There is no more crash after `taskkill /im explorer.exe`, however if you give it some time, there will be ATL assertions leading to system hang

**Review by HBelusca — APPROVED** [review]

**Pull request merged**